_posts/2023-09-19-virtual-thread-1.adoc

```diff
@@ -0,0 +1,364 @@
+
+## The Java world before Java 21
+
+At the beginning of the Java time, Java had _green threads_. 
+Green threads were user-level threads scheduled by the Java virtual machine (JVM) instead of natively by the underlying operating system (OS).
+They emulated multithreaded environments without relying on native OS abilities. 
+They were managed in user space instead of kernel space, enabling them to work in environments that do not have native thread support.
+Green threads were briefly available in Java between 1997 and 2000.
+I used green threads; they did not leave me with a fantastic memory.
+
+In Java 1.3, released in 2000, Java made a big step forward and started integrating OS threads.
+So, the threads are managed by the operating system. 
+It is still the model we are using today.
+Each time a Java application creates a thread, a platform thread is created, which wraps an OS thread.
+So, creating a platform thread creates an OS thread, and **blocking a platform thread blocks an OS thread**. 
+
+When you use a Java application framework, you rarely create threads yourself. 
+It is done for you by the framework. 
+For example, when your application receives an HTTP request, the framework creates or reuses a platform thread (and so an OS thread) and executes the processing on that thread.
+The whole processing runs on this thread, and the thread cannot be reused until the processing completes (so the response is sent back). 
+When the processing executes a blocking I/O operation, like calling another service, writing to the file system, or interacting with a database, the thread is blocked, waiting for the response. 
+As mentioned above, the OS thread is also blocked while waiting.
+When this response is received, the processing continues:
+
+image::imperative-model.png[Threads involved with the imperative model,400,float="right",align="center"]
+
+
+This model has the advantage of being simple to program with. 
+The code follows an imperative model.
+The code is executed sequentially. 
+It's simple to write, simple to reason about.
+For example, the following snippet shows how you receive an HTTP request, call another HTTP service, and return a response with Quarkus.
+It follows the sequence diagram from above.
+
+[source, java]
+----
+@Path("/greetings")
+public class ImperativeApp {
+
+  @RestClient RemoteService service;
+
+  @GET
+  public String process() {
+    // Runs on a worker (platform) thread because the
+    // method uses a synchronous signature
+
+    // `service` is a rest client, it executes an I/O operation
+    // (send an HTTP request and waits for the response)
+    var response = service.greetings(); // Blocking, it waits for the response
+                                        // The OS thread is also blocked
+
+    return response.toUpperCase();
+  }
+
+}
+----
+
+But there is a limit to that imperative model. 
+You can only handle _n_ requests concurrently, with _n_ the number of threads the framework can create.
```

Comment:
@cescoffier We can link the nice explanation using Little Law at https://youtu.be/07V08SB1l8c?t=142 

which is still valid for reactive programming still, because we enable the same thing: increased throughput by removing the dependency of the request-response latency from the number of native threads able to handle it, and by consequence, the speed of "context switch" between them.
A reactive model, similarly to loom, break this dependency and enable better performance/scalability.